Added contrib module for turn-based battle system

BattleJenkins opened the pull request:
This adds a new contrib file, a module that provides a simple framework for a tabletop style turn based battle system. From the module itself:

This is a framework for a simple turn-based combat system, similar
to those used in D&D-style tabletop role playing games. It allows
any character to start a fight in a room, at which point initiative
is rolled and a turn order is established. Each participant in combat
has a limited time to decide their action for that turn (30 seconds by
default), and combat progresses through the turn order, looping through
the participants until the fight ends.

Only simple rolls for attacking are implemented here, but this system
is easily extensible and can be used as the foundation for implementing
the rules from your turn-based tabletop game of choice or making your
own battle system.

#### Brief overview of PR changes/additions
Adds contrib module for a simple turn-based battle system.

#### Motivation for adding to Evennia
Combat systems are a near-universal feature for MUDs, and currently no contrib module provides a framework for them - this system provides a framework that resembles those of common tabletop role playing games and can be readily extended to suit the coder's needs.

#### Other info (issues closed, discussion etc)
I've tested this thoroughly on my local system and have fixed all the bugs I could find, but it hasn't been tested by multiple users yet.

Griatch commented on evennia/contrib/turnbattle.py:
```diff
@@ -0,0 +1,520 @@
+            self.caller.msg("You can't start a fight if you've been defeated!")
+            return
+        for thing in here.contents: # Test everything in the room to add it to the fight.
+            if thing.db.HP: # If the object has HP...
```
This is a nice touch, it means you could fight a table or door assuming it has HP. :)

Griatch commented on evennia/contrib/turnbattle.py:
```diff
@@ -0,0 +1,520 @@
+        Called once every self.interval seconds.
+        """
+        currentchar = self.db.fighters[self.db.turn] # Note the current character in the turn order.
+        self.db.timer -= 1 # Count down the timer by one second.
```
Ticking every second is a a simple but wasteful way to handle a countdown. The only times that the script cares about are after 20s (when the warning happens) and 30s (when the next round is force-started). You could therefore get away with ticking the script only every 10 seconds instead. You use the scripts' `restart` method whenever you want to prematurely start the next round on-demand. You can simply have your action-set commands check in to the script whenever an action is performed, so it knows when the actions are used up.

I'm not necessarily requiring you to rewrite this, but doing things on-demand is the recommended way to handle timeouts in an asynchronous system like Evennia. Checking things over and over for change is generally a bad practice.

Griatch commented on evennia/contrib/turnbattle.py:
```diff
@@ -0,0 +1,520 @@
+            return
+        
+        # Check to see if only one character is left standing. If so, end combat.
+        DefeatedCharacters = 0
```
Python PEP8 standard (and thus Evennia's style guide) dictates that local variables are always written in lowercase with underscores. CamelCase is only ever used for class names.

Griatch commented on evennia/contrib/turnbattle.py:
```diff
@@ -0,0 +1,520 @@
+----------------------------------------------------------------------------
+"""
+def roll_init(character):
+    """
```
As elsewhere, google-style docstrings are needed.

BattleJenkins commented:
Thanks a lot for all the feedback! I'm certainly still learning. I'll be able to clean most of this up on my own, but once it comes time to implement unit testing, I may end up needing some help - I've never handled unit tests before and still only have the barest idea of what they even are. In any case, I'll certainly work on making sure everything is arranged prettily and documented properly, and see what I can do about reworking the timer on the turn handler script!

Griatch commented:
@BattleJenkins The concept of a unit test is simple - they are simply example calls of the individual components of a system. making sure a given function actually returns the result you expect to a given input. These are run automatiucally and makes sure your combat system keeps working even when other things inside Evennia change. If we were to change something in the API that the combat system relies on, the unit test will hopefully fail and tell us that we must update the combat system to match. The wiki page on unit tests should explain most of what you need to know, but don't be shy to ask if you have further questions!

BattleJenkins commented:
All right, I implemented a slew of those changes and tested to make sure everything still works - let me know if there's anything I missed! I'll leave the more complicated work of re-working the turn handler script to be more efficient next.

BattleJenkins commented:
All right, I changed a bunch of stuff around and cleaned things up and I think it's worth another look! I want to make sure the module itself is in a passable state before I start on the task of learning unit tests.

Griatch commented on evennia/contrib/turnbattle.py:
```diff
@@ -0,0 +1,663 @@
+        a random integer from 1 to 1000.
+    """
+    return randint(1,1000)
+    """
```
docstrings are usually at the top of the function/method/class etc. Probably better to put this info in the Notes section of the docstring instead or, if you want to have it in the  code, make it a normal comment with # markers.

Griatch commented on evennia/contrib/turnbattle.py:
```diff
@@ -0,0 +1,663 @@
+            spend_action(currentchar, 'all', action_name="disengage") # Spend all remaining actions.
+            
+            
+    def initialize_for_combat(self, character):
```
Needs the `Args:` block.

BattleJenkins commented:
All right, those bits are fixed!

Griatch commented:
@BattleJenkins 

I ran the combat module with the latest fixes, seems to work fine. There are a few remaining issues:
1. When your turn times out, the other player only sees a message about the other timing out, there is no mention that it is now your turn. This is pretty clear on two players but could be a bit confusing for a larger fight. 
2. When trying to use an exit while in combat you get a traceback (this is a simple error, `self` on the object is the object itself, not the command):
       ```
Traceback (most recent call last):
 "/home/griatch/Devel/Home/evennia/evennia/evennia/objects/objects.py", line 683, in move_to
h     if not self.at_before_move(destination):
   File "/home/griatch/Devel/Home/evennia/evennia/evennia/contrib/turnbattle.py", line 308, in at_before_move
     self.caller.msg("You can't exit a room while in combat!")
 AttributeError: 'Character' object has no attribute 'caller'
        ```
3. It might be an idea to wrap your combat commands in a small cmdset. That way your install instructions could just be `from evennia.contrib import turnbattle` and then `self.add(turnbattle.BattleCmdSet)`.
4. I wonder if the combat should not offer some sort of custom help command. I reflexively entered help when going into combat and I figure many newbies would too - maybe help will just show a list of the available combat commands while help all and help <specific topic> will still show the normal help, something like that?

Griatch commented:
I also see some `[EE] Script Combat Turn Handler(#1) of type 'TurnHandler': at_repeat() error ''NoneType' object has no attribute 'cancel''.` errors in the logs. I wonder if this is an issue in the TurnHandler jor in the scripting mechanism itself though.

BattleJenkins commented:
I was just about to comment with that error myself - I'm looking over the code and can't seem to puzzle out why it doesn't notify a character of their turn beginning. In any case, I'm about to go to bed, so I'll have to puzzle over this more tomorrow.

BattleJenkins commented:
Commenting out line 636 fixes the issue of players not being notified of their turn, so I imagine something with the force_repeat() function isn't working properly, at least not what it's used for here - I used force_repeat to effectively reset the interval timer, because otherwise the length of turns before they time out could actually end up shorter by up to 10 seconds which wasn't a problem when the interval was only 1 second instead of 10. Which is to say, the timer might get reset to '30' when the script is only 2 seconds away from repeating, which subtracts 10 seconds from the timer on every repeat. Is there a better solution to this particular problem, aside from shortening the interval again?

BattleJenkins commented:
What the heck! Why did it fail one of the checks? It works fine on my end...

BlauFeuer commented:
The lgtm analysis doesn't seem to always run.  I don't think it's directly related to your code.  I've seen PRs accepted with the lgtm "could not build the merge commit" message, so wouldn't worry much about it.

Griatch commented:
Yeah, as Blau said, the LGTM PR analysis seems to itself sometimes fail. Travis one is the important one for stability (once the unittests are in).

BattleJenkins commented:
Ah, gotcha! Hopefully next time I make a commit, it'll check out just fine.

BattleJenkins commented:
Alrighty, that's all four issues addressed, plus one glaring oversight also fixed! With this, I think it's just about ready for prime time, once I figure out how to do unit tests.

Griatch commented:
@BattleJenkins Cool, at a quick glance it looks really nice now! I'll re-review it when the unit tests are in place. 

Btw, I turned off the LGTM Pull Request integration for now, it fails to load way too often on PRs for it to be useful.

BattleJenkins commented:
I started doing unit tests! They aren't quite so hard to figure out as I thought they would be. All the commands are covered under the unit test now.

BattleJenkins commented:
That's weird - I added the rest of the unit tests and everything's working fine for me now, but now the Travis CI build is failing. What could have gone wrong?

BattleJenkins commented:
YEAH I DID IT ALL ON MY OWN

Turns out the problem was that the turn handler script was sticking around after the unit tests were done and trying to iterate.

Griatch commented:
Looks good now, thanks for the contribution - merged with a rebase.
